server/src/main/java/org/opensearch/cluster/service/MasterService.java

```diff
@@ -131,7 +132,13 @@ public class MasterService extends AbstractLifecycleComponent {
     protected final ClusterManagerTaskThrottler clusterManagerTaskThrottler;
     private final ClusterManagerThrottlingStats throttlingStats;
 
+    private final LongSupplier nanoTimeSupplier;
+
     public MasterService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool) {
+        this(settings, clusterSettings, threadPool, System::nanoTime);
+    }
+
+    public MasterService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, LongSupplier timeProvider) {
```

Comment:
Should threadpool itself have a method to fetch non-cached time to prevent overhauling of code at multiple places. Then the cached time is 200ms by default we might want to see how system performs with lower cache times 100ms, 50ms etc and would that make sense for other use cases to switch to lower cache time or non-cached times alike

Reply:
> Should threadpool itself have a method to fetch non-cached time to prevent overhauling of code at multiple places.

+1

I dont like adding new constructor for ClusterManagerService class to just pass the timeProviderSupplier

Reply:
Seems okay, we might have to overload the ThreadPool's new time provider function for tests like we do in existing code. Will draft the changes with this suggestion.

Reply:
@Bukhtawar @shwetathareja @andrross Iterated the PR based on having a method inside ThreadPool itself to fetch absolute time. Please review now.

@shwetathareja I haven't included [ClusterApplierService](https://github.com/opensearch-project/OpenSearch/blob/aa2158555c7fea440d8de80b939a8c646aa5d415/server/src/main/java/org/opensearch/cluster/service/ClusterApplierService.java#L714) changes here now to avoid rework if further iterations are required. Will raise a separate PR once this is merged.